### ISSUE TITLE
Clang: crash on recursive RETURN_CAPABILITY attribute

### ISSUE THREAD

**Issue by klirichek** [llvm/llvm-project#60943]
MRE:
```C++
#define RETURN_CAPABILITY( x ) __attribute__ (( lock_returned ( x ) ))
#define REQUIRES( ... ) __attribute__ (( requires_capability ( __VA_ARGS__ ) ))

using SchedRole [[clang::capability("role")]] = bool;

class foo {
    SchedRole m_thread;
public:
    // notice - erroneously return cap getRole() instead of m_thread here.
	SchedRole getRole() const RETURN_CAPABILITY ( getRole() )
	{
		return m_thread;
	}
};

void bar (foo& a) REQUIRES(a.getRole()) {}
```

Need to be compiled with `-Wthread-safety`

Most of the versions just silently returns with code 139, however I've found one which produces readable backtrace.

https://godbolt.org/z/766v79nh1

**Comment by shafik** [comment]
Confirmed, I don't really know what the semantics should be here but we should not be crashing. If this is ill-formed we should issue a diagnostic.

**Comment by TheInferentialImp** [comment]
It's ill-formed I believe. The code defines two macros: RETURN_CAPABILITY and REQUIRES, which are used to annotate functions and parameters with capabilities. The SchedRole type is defined with a [[clang::capability]] attribute.

The foo class has a private member m_thread of type SchedRole. The getRole() function returns this member, but it is erroneously annotated with RETURN_CAPABILITY(getRole()), which means that the function returns a lock for the getRole() capability, instead of the m_thread member. This is likely a mistake.

The bar() function takes a reference to a foo object and is annotated with REQUIRES(a.getRole()), which means that it requires the a object to have the getRole() capability. This is used to ensure that the function is only called with objects that have the required capability.

**Comment by aaronpuchert** [comment]
This basically says that `getRole()` returns `getRole()`, which ends up producing a stack overflow in evaluating the substitution. It might get tricky to find all such recursions because we currently still allow adding attributes on redeclarations, see #42194.

But we could produce a warning/error on direct self-reference, i.e. if the attribute references the declaration it is attached to. @AaronBallman, what do you think? (We could then declare the attribute as invalid, which should prevent the stack overflow.)

> we should not be crashing

“Crash” is a platform-specific behavior. Memory safety violations or assertion failures should always be fixed, but stack overflow is resource exhaustion, which can happen for pathological inputs. This should be fixed because it's infinite recursion (which is UB).

**Comment by AaronBallman** [comment]
> This basically says that `getRole()` returns `getRole()`, which ends up producing a stack overflow in evaluating the substitution. It might get tricky to find all such recursions because we currently still allow adding attributes on redeclarations, see #42194.
> 
> But we could produce a warning/error on direct self-reference, i.e. if the attribute references the declaration it is attached to. @AaronBallman, what do you think? (We could then declare the attribute as invalid, which should prevent the stack overflow.)

Catching at least the direct self-reference case seems important. I agree that this might be difficult to do in general, but making efforts to catch what we can makes sense to me.

> > we should not be crashing
> 
> “Crash” is a platform-specific behavior. Memory safety violations or assertion failures should always be fixed, but stack overflow is resource exhaustion, which can happen for pathological inputs. This should be fixed because it's infinite recursion (which is UB).

Even for pathological inputs, we should not be causing a fault. We've got some utilities to try to help mitigate stack exhaustion: https://github.com/llvm/llvm-project/blob/80fe3168b507587cd9db47115de609d98188c6f5/clang/lib/Sema/Sema.cpp#L511
